### DISCUSSION TITLE
Is it possible to save the whole message in a local file?

### DISCUSSION THREAD

**Discussion by alexv96** [php-mime-mail-parser/php-mime-mail-parser#372] [Q&A] [answered]
Hi, I wonder if there is a function or method in the library to save all the email I capture either in an .eml file or another.

**Comment by eXorus** [comment]
My library parse an .eml file so I don't really see the point to have a method to get the file because the file was given by the user.

**Reply by alexv96** [reply]
> My library parse an .eml file so I don't really see the point to have a method to get the file because the file was given by the user.

 Or a way to save the whole message if it is to have it as a backup in case there is a bad process in the application I have. 

Since we are working with procmail.

For the moment I managed to save the complete header in case so, if the idea is to have only a complete copy of the message or of the important header and body of the message to then execute the command by hand and correct the error.

```
$nameFile = $website->hostnames->implode('fqdn') .'_'. time().'.eml';
$createFile = fopen($nameFile,'w+');
fwrite($createFile,$parser->getHeadersRaw());
fwrite($createFile,print_r(['body'=>$parser->getMessageBody('text')],true));
fclose($createFile);
```

**Reply by alexv96** [reply]
> My library parse an .eml file so I don't really see the point to have a method to get the file because the file was given by the user.

I managed to capture the message and not only the Header in RAW while I have it in hard in the library which is to change only one word.

using this getHeader() method that returns in RAW format and I had to change in these 2 getPartHeaderFromText() getPartHeaderFromFile() methods the following line

`
    protected function getPartHeaderFromFile(&$part)
    {
        $start = $part['starting-pos'];
        $end = $part['ending-pos-body'];// $part['starting-pos-body'];
        fseek($this->stream, $start, SEEK_SET);
        $header = fread($this->stream, $end - $start);
        return $header;
    }

    /**
     * Retrieve the Header from a MIME part from text
     *
     * @return String Mime Header Part
     * @param $part Array
     */
    protected function getPartHeaderFromText(&$part)
    {
        $start = $part['starting-pos'];
        $end = $part['ending-pos-body'];// $part['starting-pos-body']; //CHANGE WORD STARTING BY ENDING
        $header = substr($this->data, $start, $end - $start);
        return $header;
    }`

**Reply by eXorus** [reply]
In my other project mailcare, the command that receive the email from postfix but I guess it's the same with procmail I do:

```
$fd = fopen("php://stdin", "r");
$rawEmail = "";
while (!feof($fd)) {
    $rawEmail .= fread($fd, 1024);
}
fclose($fd);
```

**Reply by aiso-net** [reply]
I would like the same, i'm trying to do what altermime does (so old it doesnt like current email formats), parse a raw file passed via stdin, add a prepended and/or appended disclaimer to the plain and html part of the message and then save the complete modified message with attachments back out to a file/stdout

**Reply by eXorus** [reply]
What you are looking for is a way to get an email (parse it so my lib) and create a new one (not my lib more symfony/mailer).

I don't really know how to do it, because I know how to decode an email but I don't have any method to rebuild it.

**Reply by aiso-net** [reply]
Ya, the high level way looks like taking your decoded results, editing them and then plugging the edited results into PHPMailer and then doing in PHPMailer:
if (!$mail->preSend()) {
    echo $mail->ErrorInfo;
} else {
    $message = $mail->getSentMIMEMessage();
}
This would return a raw complete RFC822 message but I dont know the exact coding to take what you class provides and plug it into PHPMailers functions.

**Reply by eXorus** [reply, marked as answer]
Yes I know it's difficult but I know someone already did it for https://anonaddy.com/.

Forwarding an email is easy => Mail Server
Get emails and parse it => MailServer + php-mime-mail-parser
Send emails => MailServer + SF Mailer
Get emails, change it and sent it => ????

**Reply by aiso-net** [reply]
Thanks, I was able to handle it using your library and PHPMailer and a bunch of code. I now made a PHP replacement of altermime for disclaimer adding and works great
